libraries/chain/db_market.cpp

```diff
@@ -1293,4 +1331,31 @@ asset database::pay_market_fees(const account_object* seller, const asset_object
    return market_fees;
 }
 
+/***
+ * @brief Calculate force-settlement fee and give it to issuer of the settled asset
+ * @param collecting_asset the smart asset object which should receive the the fee
+ * @param collat_receives the amount of collateral the settler would expect to receive absent this fee
+ *     (fee is computed as a percentage of this amount)
+ * @return asset denoting the amount of fee collected
+ */
+asset database::pay_force_settle_fees(const asset_object& collecting_asset, const asset& collat_receives)
+{
+   const asset_object& recv_asset = get(collat_receives.asset_id);
+   FC_ASSERT( collecting_asset.get_id() != collat_receives.asset_id );
+   if( !collecting_asset.options.extensions.value.force_settle_fee_percent.valid()
+         || *collecting_asset.options.extensions.value.force_settle_fee_percent == 0 )
+      return recv_asset.amount(0);
+   auto value = detail::calculate_percent(collat_receives.amount, *collecting_asset.options.extensions.value.force_settle_fee_percent);
+   asset settle_fee = recv_asset.amount(value);
+   // Deposit fee in asset's dynamic data object:
+   if( value > 0)
+   {
+      const auto& asset_dyn_data = collecting_asset.dynamic_asset_data_id(*this);
+      modify( asset_dyn_data, [&settle_fee]( asset_dynamic_data_object& obj ){
+         obj.accumulated_fees += settle_fee.amount;  // TODO: Wrong place to put fee
```

Comment:
Yes, I'm aware of it.

Please keep in mind that the collateral asset type of a debt asset can be changed, so we probably need a multi-asset container for storing the fee. Another option is, only allows the collateral asset type to be changed when the container is empty.

BSIP74 needs the container as well. Please coordinate with @jmjatlanta to avoid duplicate efforts.

@dls-cipher please keep an eye on this matter.

Reply:
@abitmore - please confirm that BSIP74 and BSIP87 have diff container types.

Thanks

Reply:
@dls-cipher I think bsip74 and bsip87 can use the same container.